### DISCUSSION TITLE
Extra assignment

### DISCUSSION THREAD

**Discussion by n9** [JuliaLang/julia#44188] [General]
@mgkuhn

https://github.com/JuliaLang/julia/blob/cc345f6bc5f1406bd8d5af8265b91645f0adaeaf/base/strings/io.jl#L622

Is this assignment needed?

Update: I have just found a closed PR for this: https://github.com/JuliaLang/julia/pull/38499 
I can't see the build details to see why it failed. (It shows empty screen.)

**Comment by mgkuhn** [comment]
No, that line can be deleted from `escape_raw_string`.

 It probably was a leftover from a previous draft of the algorithm, e.g. one that may have used `write(io, '\\' ^ escapes)` instead of `while escapes > 0; write(io, '\\'); escapes -= 1; end`. Given Julia's use of unbuffered standard output #43176, I'm still not sure in such situations whether I should minimize the number of allocations or the number of `write` calls and methods used.

**Comment by n9** [comment]
Thank you for your answer. This will also answer my second question, whether this algorithm is currently equivalent to:

```jl
function escape_raw_string(io, str::AbstractString)
    for c in str
        if c in ('\\', '"')
            write(io, '\\')
        end
        write(io, c)
    end
end
```

If yes, I was wondering, why the code is currently complicated.

**Reply by mgkuhn** [reply]
That's a different algorithm that is incompatible with how Julia raw string escaping works.

**Reply by mgkuhn** [reply]
It's more like how POSIX shell escaping works rather than how Microsoft C escaping command-line parameter escaping works. (In other words, it is more sane. Julia raw strings went with the less sane Microsoft C approach, which is not a homomorphism under string concatenation.)

**Reply by n9** [reply]
Ok. I have to download Julia and try them both:) My in-brain running skill seems to fail.

**Reply by n9** [reply]
Could you please provide me a case where both algorithms return different results? (I am unable to find one.)

**Reply by n9** [reply]
... Nevermind, I have found one: `"c:\\Program Files\\"`.

**Comment by mgkuhn** [comment]
Since we are looking at `escape_raw_string` again: one thing I had on my todo list was to extend this function (e.g., with two optional arguments `escapes = Ref{Int}(0), last = true`) such that it can be used in streaming mode. If you have a multi-gigabyte string that you want to escape into a raw string, you may want to do this without holding the entire string in memory, processing just chunks at a time. To enable this, we'd need:
* a way to preserve the `escapes` value across invocations (perhaps by making it a reference argument?)
* a way to tell the function if it is processing the `last` chunk in a sequence of string chunks (because after the last chunk, it must reduce `escapes` to zero)

So when I make a PR for that, I'll also delete that line.

**Comment by n9** [comment]
I have one more question regarding this function. Why this function is not used around `dir` in `pushd`?

https://github.com/JuliaLang/julia/blob/cc345f6bc5f1406bd8d5af8265b91645f0adaeaf/stdlib/Distributed/src/managers.jl#L312

**Reply by mgkuhn** [reply]
If I remember correctly, Windows `pushd` does not understand that escaping convention, and cannot deal with paths that contain a `"`. Try it out.

**Reply by mgkuhn** [reply]
[Windows file names](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) are not allowed to contain `"` characters. And without `"` characters there is nothing to escape here.

**Reply by n9** [reply]
Make sense, thank you! (Maybe the surrounding quotes are also not needed there.)

**Reply by mgkuhn** [reply]
These quotes are needed by `cmd.exe` to disable other meta characters that might occur in the filename (such as the `&&` that follows). I've tested this routine quite carefully to correctly treat all allowed characters in Windows filenames.

**Reply by mgkuhn** [reply]
If you think I got something wrong, please provide an example.

**Reply by mgkuhn** [reply]
In case you are trying to understand this code to port it elsewhere: my advice is to keep in mind that Windows command-line syntax is wild and crazy, and that nothing you might expect from being familiar with POSIX shells will apply. After a lot of time, reading and experimentation, I think I have mastered most of `cmd.exe` syntax, but I am still really struggling to e.g. make sense of PowerShell syntax, which seems even worse ...

The main reason why I still haven't finished the PowerShell option of the `launch_on_machine` function is that the safest option I found so far is to invoke `cmd.exe` from PowerShell. (Widely used PowerShell versions have no equivalent of `&&`.)

**Reply by n9** [reply]
> In case you are trying to understand this code to port it elsewhere ...

I am just trying to invoke Windows executable via SSH command (noninteractively).

> ... keep in mind that Windows command-line syntax is wild and crazy ...

Yes, currently, the only documentation I have found is this: https://docs.microsoft.com/en-us/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way

> ... but I am still really struggling to e.g. make sense of PowerShell syntax, which seems even worse ...

I came here from https://github.com/PowerShell/Win32-OpenSSH/issues/1082.
